source/CIS/cis-5.rst

```diff
@@ -0,0 +1,599 @@
+
+``WithdrawCis2TokensEvent``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``WithdrawCis2TokensEvent`` SHALL be logged for every ``withdrawCis2Tokens`` function invoke.
+
+The ``WithdrawCis2TokensEvent`` is serialized as: First a byte with the value of 246, followed by the
+:ref:`CIS-5-TokenAmount` (``tokenAmount``), :ref:`CIS-5-TokenID` (``TokenID``),
+:ref:`CIS-5-ContractAddress` (``cis2TokenContractAddress``), a :ref:`CIS-5-PublicKeyEd25519` (``from``), and the :ref:`CIS-5-Address` (``to``)::
+
+  WithdrawCis2TokensEvent ::= (246: Byte) (tokenAmount: TokenAmount) (tokenId: TokenID) (cis2TokenContractAddress: ContractAddress) (from: PublicKeyEd25519) (to: Address)
+
+``InternalNativeCurrencyTransferEvent``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``InternalNativeCurrencyTransferEvent`` SHALL be logged for every ``internalNativeCurrencyTransfer`` function invoke.
+
+The ``InternalNativeCurrencyTransferEvent`` is serialized as: First a byte with the value of 245, followed by the :ref:`CIS-5-CCDAmount` (``ccdAmount``), a :ref:`CIS-5-PublicKeyEd25519` (``from``), and the :ref:`CIS-5-PublicKeyEd25519` (``to``)::
+
+  InternalNativeCurrencyTransferEvent ::= (245: Byte) (ccdAmount: CCDAmount) (from: PublicKeyEd25519) (to: PublicKeyEd25519)
+
+``InternalCis2TokensTransferEvent``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``InternalCis2TokensTransferEvent`` SHALL be logged for every ``internalCis2TokensTransfer`` function invoke.
+
+The ``InternalCis2TokensTransferEvent`` is serialized as: First a byte with the value of 244, followed by the
+:ref:`CIS-5-TokenAmount` (``tokenAmount``), :ref:`CIS-5-TokenID` (``TokenID``),
+:ref:`CIS-5-ContractAddress` (``cis2TokenContractAddress``), a :ref:`CIS-5-PublicKeyEd25519` (``from``), and the :ref:`CIS-5-PublicKeyEd25519` (``to``)::
+
+  InternalCis2TokensTransferEvent ::= (244: Byte) (tokenAmount: TokenAmount) (tokenId: TokenID) (cis2TokenContractAddress: ContractAddress) (from: PublicKeyEd25519) (to: PublicKeyEd25519)
+
+
+
+.. _CIS-5-functions:
+
+Contract functions
+------------------
+
+A smart contract implementing this standard MUST export the following functions:
+
+- :ref:`CIS-5-functions-depositNativeCurrency`
+- :ref:`CIS-5-functions-depositCis2Tokens`
+- :ref:`CIS-5-functions-withdrawNativeCurrency`
+- :ref:`CIS-5-functions-withdrawCis2Tokens`
+- :ref:`CIS-5-functions-internalNativeCurrencyTransfer`
+- :ref:`CIS-5-functions-internalCis2TokensTransfer`
+- :ref:`CIS-5-functions-balanceOfNativeCurrency`
+- :ref:`CIS-5-functions-balanceOfCis2Tokens`
+
+
+.. _CIS-5-functions-depositNativeCurrency:
+
+``depositNativeCurrency``
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The function is payable and deposits/assigns the send CCDAmount (native currency) to a public key (``PublicKeyEd25519``).
+
+Parameter
+~~~~~~~~~
+
+The parameter is a ``PublicKeyEd25519``.
+
+See the serialization rules in :ref:`CIS-5-PublicKeyEd25519`.
+
+Requirements
+~~~~~~~~~~~~
+
+- The function MUST emit a ``DepositNativeCurrencyEvent``.
```

Comment:
_Tied to my comment above related to requirements for the events._

I don't think the event should be used to signal the function call itself, but rather the _event_ of CCD being deposited. So I don't think this requirement is needed.